diesel/src/pg/metadata_lookup.rs

```diff
@@ -9,32 +9,25 @@ use std::cell::RefCell;
 use std::collections::HashMap;
 
 /// Determines the OID of types at runtime
-#[allow(missing_debug_implementations)]
-#[repr(transparent)]
-pub struct PgMetadataLookup {
-    conn: PgConnection,
-}
-
-impl PgMetadataLookup {
-    #[allow(clippy::new_ret_no_self)]
-    pub(crate) fn new(conn: &PgConnection) -> &Self {
-        unsafe { &*(conn as *const PgConnection as *const PgMetadataLookup) }
-    }
-
+pub trait PgMetadataLookup {
     /// Determine the type metadata for the given `type_name`
     ///
     /// This function should only be used for user defined types, or types which
     /// come from an extension. This function may perform a SQL query to look
     /// up the type. For built-in types, a static OID should be preferred.
-    pub fn lookup_type(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
-        let metadata_cache = self.conn.get_metadata_cache();
+    fn lookup_type(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata;
+}
+
+impl PgMetadataLookup for PgConnection {
+    fn lookup_type(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
```

Comment:
I think I would prefer an implementation where the caching done here is not something connection specific, but the lookup interface encourages all implementations to do a similar kind of caching. Otherwise this will likely cause performance issues with custom connection implementations.

Now I do not have a clear opinion here how exactly the interface should look like. Probably this can be implemented by some default functions available on the `PgMetadataLookup` trait + a getter for the cache?

Reply:
Looks like this is solved by the interface proposed at https://github.com/diesel-rs/diesel/pull/2639#discussion_r570487328

Reply:
To solve this issue on the above interface wouldn't you have to store at least a reference to the cache on `PgMetadataLookup`.

Reply:
Not really, for example `PgMetadataLookup` could look somehow like this:

```rust
trait PgMetadataLookup {
    fn get_cache_map(&self) -> PgMetadataCache
    fn lookup_type(&self) -> PgTypeMetadata { // a default impl here that works for all connections that have `Backend = Pg`}
}
```

Reply:
Ok, I will start work on this.